### PR TITLE
Categories system for parties

### PR DESCRIPTION
**Pull request by LordMidas**
This allows specifying categories within parties where each category is a party. 

Just like #7 it works. I haven't "beautified" the code or updated most comments yet but it is functional. I'd like to merge this so that going forward we can base off of this to make further improvements.

### CODE REVIEW

File: mod_dynamic_spawns/classes/party.nut [hunk 1]
```diff
@@ -55,6 +77,17 @@
 			}
 		}
 
+		this.__SubParties = [];
```
Comment (Darxo) [inline mod_dynamic_spawns/classes/party.nut:80]: I'm confused that you call this field `__SubParties` when it is created purely from categories.
What about `__Categories` instead?

File: mod_dynamic_spawns/classes/party.nut [hunk 2]
```diff
@@ -144,39 +180,118 @@
...
 	// Checks only against HardMax and HardMin
-	function canSpawn( _spawnProcess )
+	function canSpawn()
 	{
-		if (_spawnProcess.getTotal() >= this.getHardMax()) return false;
+		if (this.__SpawnProcess.getTotal() >= this.getHardMax()) return false;
+
+		// RatioMax is ignored if we do not satisfy the RatioMin yet
+		if (!this.satisfiesRatioMin())
+				return true;
```
Comment (Darxo) [inline mod_dynamic_spawns/classes/party.nut:189]: weird double indent

File: mod_dynamic_spawns/classes/spawn_process.nut [hunk 3]
```diff
@@ -63,115 +80,171 @@
...
+	}
+
+	function getSpawnables()
+	{
+		return this.__SubSpawnProcesses.len() == 0 ? this.getParty().getUnitBlocks() : this.getParty().getSubParties();
+	}
+
+	function getSpawnable( _id )
+	{
+		local spawnables = this.__SubSpawnProcesses.len() == 0 ? this.getParty().getUnitBlocks() : this.getParty().getSubParties();
```
Comment (Darxo) [inline mod_dynamic_spawns/classes/spawn_process.nut:115]: this should make use of the getSpawnables() function defined directly above

File: mod_dynamic_spawns/classes/party.nut [hunk 4]
```diff
@@ -144,39 +180,118 @@
...
+	function onSpawnEnd()
+	{
+		foreach (unitBlock in this.getUnitBlocks())
+		{
+			unitBlock.onSpawnEnd();
+		}
+	}
+
+	// Returns true if the ratio of this unitblock would still be below its defined RatioMax if it was to spawn the next unit
+	// _spawnProcess = current spawnprocess reference that includes most important variables
```
Comment (Darxo) [inline mod_dynamic_spawns/classes/party.nut:235]: this comment is now redundant

File: mod_dynamic_spawns/classes/party.nut [hunk 5]
```diff
@@ -144,39 +180,118 @@
...
+		if (mainSpawnProcess == null)
+			return true;
+
+		local referencedTotal = ::Math.max(mainSpawnProcess.getTotal() + 1.0, mainSpawnProcess.getParty().getHardMin());
+		local maxAllowed = ::Math.round(this.RatioMax * referencedTotal);
+		return (mainSpawnProcess.getCategoryTotal(this.getID()) < maxAllowed);
+	}
+
+	// Returns true if the ratio of this unitblock would still be above its defined RatioMin if it was to spawn the next unit
+	// _spawnProcess = current spawnprocess reference that includes most important variables
```
Comment (Darxo) [inline mod_dynamic_spawns/classes/party.nut:248]: this comment is now redundant

File: mod_dynamic_spawns/classes/party.nut [hunk 6]
```diff
@@ -144,39 +180,118 @@
 
 	// Returns false if there is a reason why this party is not allowed to spawn anymore. This function only checks for a small subset of all possible reasons.
 	// Checks only against HardMax and HardMin
```
Comment (Darxo) [inline mod_dynamic_spawns/classes/party.nut:182]: this line is no longer correct and should be removed

**Review by Darxo — APPROVED** [review]
Looks ok

All those changes are somewhat confusing to me.
Could me lack of comments, the fact that I wasn't the one writing them or just the sheer amount of changes taken in all at once.

A small comment or explanation is enough to close the issues I raised during this review if you disagree with my concerns.

File: mod_dynamic_spawns/classes/party.nut (hunk 1 above)
Comment (LordMidas) [reply mod_dynamic_spawns/classes/party.nut:80]: `Categories` is a frontend variable used by "designers" to design parties in code. Whereas in fact they are actual party objects. So the private variable used here is called `SubParties` to make it clear in code that these are party objects and I prefer this.

**Comment by LordMidas** [comment]
> Looks ok
> 
> All those changes are somewhat confusing to me. Could me lack of comments, the fact that I wasn't the one writing them or just the sheer amount of changes taken in all at once.
> 
> A small comment or explanation is enough to close the issues I raised during this review if you disagree with my concerns.

Basically major changes include:
- Adding Categories as a frontend variable for partydefs. This is converted party objects which are stored in `__SubParties` on the inside.
- Spawn processes can now have sub spawn processes (these handle the spawning of subparties aka categories).
- Removed the passing around of `_spawnProcess` as a parameter and each spawnable has a private weak reference to its relevant spawn process.

I've updated the comments and other changes you highlighted.

**Pull request merged**